base/applications/rapps/include/asyncinet.h

```diff
@@ -0,0 +1,68 @@
+#ifndef ASYNC_INET
+#define ASYNC_INET
+
+#include <windows.h>
+#include <wininet.h>
+
+enum ASYNC_EVENT
+{
+    ASYNCINET_DATA,           // wParam is the Data retrieved from the internet, lParam is the length of Data
+
+    ASYNCINET_COMPLETE,       // wParam and lParam is not used.
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+
+    ASYNCINET_CANCELLED,      // wParam and lParam is not used.
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+
+    ASYNCINET_ERROR           // wParam is not used. lParam specify the error code (if there is one). 
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+};
+
+typedef struct __AsyncInet ASYNCINET, * pASYNCINET;
+
+typedef int
+(*ASYNCINET_CALLBACK)(
+    pASYNCINET AsyncInet,
+    ASYNC_EVENT Event,
+    WPARAM wParam,
+    LPARAM lParam,
+    VOID* Extension
+    );
+
+typedef struct __AsyncInet
+{
+    HINTERNET hInternet;
+    HINTERNET hInetFile;
+
+    HANDLE hEventHandleCreated;
+
+    UINT ReferenceCnt;
+    CRITICAL_SECTION CriticalSection;
+    HANDLE hEventRefZero; // TODO: Use conditional variable instead when one day ROS has conditional variable
+    HANDLE hEventHandleClose;
```

Comment:
If you use reference counting correctly, you would not need those handles, and probably also not the critical section.
Basically:

Before doing anything, a thread (or function) increases reference count.
When starting a thread or sending something to another thread, (thread a creates thread b):

1. Thread a increments reference count
2. Thread a starts thread b
3. Thread b is responsible of decrementing the reference count when it is done with the object.